File: test/lanetrack.js

function lanetrack(cv, frame) {
    let i = cv.imread(frame);
    let gray = new cv.Mat();
    cv.cvtColor(i, gray, cv.COLOR_BGR2GRAY) 

    cv.threshold(gray, gray, 100, 255, cv.THRESH_BINARY)


    let M = cv.Mat.ones(5, 5, cv.CV_8U);
    cv.dilate(gray, gray, M, new cv.Point(-1, -1), 2, cv.BORDER_CONSTANT, cv.morphologyDefaultBorderValue())


    cv.erode(gray, gray, M, new cv.Point(-1, -1), 2, cv.BORDER_CONSTANT, cv.morphologyDefaultBorderValue())


    img_row = gray.rows
    img_col = gray.cols
    dst_row = Math.ceil(img_row * 0.9)
    console.log("dst_row %d", dst_row)

    let color = []
    dst_col_start = Math.ceil(img_col * 0.2)
    dst_col_end = Math.ceil(img_col * 0.8)
    for (let i = dst_col_start; i < dst_col_end; i++) {
        //color.push(gray.data[dst_row * img_col + i])
        color.push(gray.ucharAt(dst_row, i))
    }
    console.log(color)

    let white_count = 0
    let white_index = []
    for (let i = 0; i < color.length; i++) {
        if (color[i] == 255) {
            white_count++,
                white_index.push(i)
        }
    }
    console.log(white_count)
    i.delete();
    gray.delete();
    M.delete();
    if (white_count > 0) {
        var center = (white_index[white_count - 1] + white_index[0]) / 2 + dst_col_start
        console.log("center %d", center)


        return center

    } else {
        return -1
    }
}

File: test/LaneTrack.py
import cv2
import numpy as np
import time


def detect(frame):
    gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
    retval, dst = cv2.threshold(gray, 100, 255, cv2.THRESH_BINARY)
    dst = cv2.dilate(dst, None, iterations=2)
    dst = cv2.erode(dst, None, iterations=6)
    # 单看第400行的像素值v
    color = dst[400]  # 取图像下1/4部为小车行驶方向判断点
    # 找到白色的像素点个数，如寻黑色，则改为0，白色为255
    white_count = np.sum(color == 255)
    # 找到白色的像素点索引，如寻黑色，则改为0，白色为255
    white_index = np.where(color == 255)
    # 防止white_count=0的报错
    if white_count == 0:
        white_count = 1
    # 找到黑色像素的中心点位置
    # 计算方法应该是边缘检测，计算白色边缘的位置和/2，即是白色的中央位置。
    center = (white_index[0][white_count - 1] + white_index[0][0]) / 2

    return center
